File: src/test/java/io/trydent/olimpo/bus/CommandProcessTest.java
package io.trydent.olimpo.bus;

import io.vertx.core.eventbus.EventBus;
import io.vertx.core.eventbus.MessageConsumer;
import io.vertx.core.json.JsonObject;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static io.trydent.olimpo.sys.Id.id;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

class CommandProcessTest {
  private interface JsonMessageConsumer extends MessageConsumer<JsonObject> {}

  private final EventBus eventBus = mock(EventBus.class);
  private final MessageConsumer<JsonObject> consumer = mock(JsonMessageConsumer.class);

  @Test
  @DisplayName("should execute a command")
  void shouldExecuteCommand() {
    when(eventBus.<JsonObject>localConsumer(eq("any-command"), any())).thenReturn(consumer);

    final var id = id("process-id");
    final var process = new CommandProcess(
      id,
      eventBus
    );

    process.accept("any-command", params -> {});

    verify(eventBus).<JsonObject>localConsumer(eq("any-command"), any());
  }
}

File: src/test/java/io/trydent/olimpo/bus/CommandProcessTest.kt
package io.trydent.olimpo.bus

import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import io.trydent.olimpo.sys.Id.Companion.id
import io.trydent.olimpo.vertx.json.Json
import io.vertx.core.eventbus.EventBus
import io.vertx.core.eventbus.MessageConsumer
import org.junit.jupiter.api.Test

internal class CommandProcessTest {
  private val eventBus: EventBus = mockk()
  private val consumer: MessageConsumer<Json> = mockk()

  @Test
  internal fun `should execute a command`() {
    every { eventBus.localConsumer<Json>("any-command", any()) } returns consumer

    val id = id("process-id")
    val process = CommandProcess(
      id,
      eventBus
    )

    process("any-command", execution = object : Execution {
      override fun invoke(json: Json) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
      }
    })

    verify {
      eventBus.localConsumer<Json>("any-command") {
        it.apply {
          reply(
            json(
              "processId" to id()
            )
          )
        }
      }
    }
  }
}
